### PR TITLE
Rating system

### PR DESCRIPTION
**Pull request by Trimardio**
the glicko system works and the basic ui works too.
Still in dev but should be usable for tests of the glicko system

be careful when entering games, you have to use the precise name of the player, case-sensitive already used previously notably, else it'll create a second player 

Not yet updated with main, please do not merge.

### CODE REVIEW

**Comment by frankystone** [comment]
Maybe i am wrong but i think the diff is not complete: E.g. the file wlrating/migrations/0001_initial.py is not shown completely?

Please add the models field option 'on_delete' for ForeignField, see https://docs.djangoproject.com/en/1.11/ref/models/fields/#django.db.models.ForeignKey.on_delete. This is mandatory for the upgrade to a later django version.

Don't know when i find the time to review and test this pull request.

**Comment by Trimardio** [comment]
Please don't merge yet, I made the pull request so people can easily test it. But I didn't even update with the last version of main.

Noted point on the 'on_delete' feature, will see about that.

As for why the file isn't showed completly that is because I make this pull request for the branch rating_system. Hence, this file was already present, and I made very little update since this branch was present in bzr

**Comment by gunchleoc** [comment]
Maybe create a linked issue for this, and discuss any needed changes in the issue before making a pull request?

Pull requests on the website don't trigger anything, and we can always get a full diff on GitHub without needing to send of a pull request. One of the advantages over GitHub :)

File: wlrating/view.py [hunk 1]
```diff
@@ -177,6 +236,10 @@ def type_to_int(word):
         return 2
     if word == 'collectors':
         return 3
+    if word == 'territorial lord':
+        return 4
+    if word == 'artifacts':
+        return 5
```
Comment (blotter) [inline wlrating/view.py]: bitte als array/dict ab in die settings.py und sortieren oder bei gleichem return mit if ... or ... or ...:

File: wlrating/view.py [hunk 2]
```diff
@@ -136,30 +173,52 @@ def score (request):
...
+            if "team" in dict_property:
+                num = str(dict_property[-1])
+                if not num in player_list:
+                    player_list[num] = {} 
+                player_list[num]['team'] = value
+                print (player_list[num])
+
+    data = []
+
+    print (player_list)
```
Comment (blotter) [inline wlrating/view.py]: ???

File: wlrating/view.py [hunk 3]
```diff
@@ -54,67 +54,105 @@ def arbiter (request):
...
+
+    for pr in Player_Rating.objects.all():
+        pr.delete()
 
-    #To move to separate section so we don't recalculte the score every time someone look at them
     for tu in Temporary_user.objects.all():
         nb_of_games = 0
         win= 0
-        
+        print (tu.username)
```
Comment (blotter) [inline wlrating/view.py]: ???

File: wlrating/templates/wlrating/arbiter.html [hunk 4]
```diff
@@ -7,49 +7,95 @@
...
+<script src="{% static 'js/arbiter.js' %}" type="text/javascript"></script>
+<script type="text/javascript">
+    document.addEventListener('DOMContentLoaded', function(){ 
+        setAddPlayerBtn()
+    }, false);
+</script>
+
 <div class="blogEntry post">
-    <table id="arbiter-table">
+    <table id="arbiter-table"  cellspacing="0" cellpadding="0">
```
Comment (blotter) [inline wlrating/templates/wlrating/arbiter.html]: ab ins css damit -> https://stackoverflow.com/questions/339923/set-cellpadding-and-cellspacing-in-css

File: wlrating/glicko2.py [hunk 5]
```diff
@@ -8,149 +8,233 @@
...
 
-        if Delta**Decimal(2) > φ**Decimal(2) + v:
-            B = math.log(Delta**Decimal(2) - φ**Decimal(2) - v)
+        if Delta**Decimal(2) > phi**Decimal(2) + nu:
+            B = Decimal(math.log(Delta**Decimal(2) - phi**Decimal(2) - nu))
         else:
             k = Decimal(1)
-            while self.function_f(a - Decimal(k)*Decimal(self.volatility_change), Delta, φ, v, σ):
+            iteration_num = 0
+            while self.step5_function_f(a - Decimal(k)*Decimal(self.tau), Delta, phi, nu, sigma) and iteration_num < self.iteration_limit:
```
Comment (blotter) [inline wlrating/glicko2.py]: Zeilenumbruch nach maximal 100 Zeichen

File: wlrating/glicko2.py [hunk 6]
```diff
@@ -8,149 +8,233 @@
...
+            games_until_last -= 1
+            print (games_until_last)
+            if games_until_last == 0:
+                self.calculate_round(games, s)
+            g.counted_in_score= True
+            g.save()
+
+
+    def calculate_round(self, games, s):
+        print ("NEW ROUND")
```
Comment (blotter) [inline wlrating/glicko2.py]: ???

File: wlrating/glicko2.py [hunk 7]
```diff
@@ -8,149 +8,233 @@
...
         for g in games:
             print (g.start_date)
-        data = self.get_temp_data(games)
+            print (g.game_map)
```
Comment (blotter) [inline wlrating/glicko2.py]: ???

File: wlrating/glicko2.py [hunk 8]
```diff
@@ -8,149 +8,233 @@
...
-            pr.decimal3 = self.volatility
+            pr = Player_Rating.objects.get(user=tu, rating_type= 3)
+            pr.decimal1 = rating
+            pr.decimal2 = deviation
+            pr.decimal3 = volatility
             pr.save()
 
+            
+    def glicko_process(self, player):
+        print (player['username'])
```
Comment (blotter) [inline wlrating/glicko2.py]: ???

File: wlrating/glicko2.py [hunk 9]
```diff
@@ -8,149 +8,233 @@
...
-        r = Decimal(173.7178) * μ + Decimal(1500)
-        return r
-    
-    def stp8_deviation(self, φ):
-        RD = Decimal(173.7178) * φ
-        return RD
+        ## Step 5
+        sigma_prime = self.step5_sigma(Delta, phi, nu, sigma)
+        print (sigma_prime)
+        print ('yOOOOO?')
```
Comment (blotter) [inline wlrating/glicko2.py]: ???

File: wlrating/glicko2.py [hunk 10]
```diff
@@ -8,149 +8,233 @@
...
+        ## Step 7
+        phi_prime = self.step7_phi_prime(phi_star,nu)
+        mu_prime = self.step7_mu_prime(player['games'], mu, phi_prime)
+
+        ## Step 8
+        final_score = self.step8_rating(mu_prime)
+        final_deviation = self.step8_deviation(phi_prime)
+        final_volatility = sigma_prime
+
+        print ('final_volatility', final_volatility)
```
Comment (blotter) [inline wlrating/glicko2.py]: ???

File: wlrating/glicko2.py [hunk 11]
```diff
@@ -8,149 +8,233 @@
...
+            if round_games < self.game_per_round:
+                games.append(g)
+                round_games+=1
+            else:
+                games.append(g)
+                self.calculate_round(games, s)
+                round_games = 0
+                games = []
+            games_until_last -= 1
+            print (games_until_last)
```
Comment (blotter) [inline wlrating/glicko2.py]: ???

File: wlrating/glicko2.py [hunk 12]
```diff
@@ -8,149 +8,233 @@
...
         self.epsilon = EPISLON
+        self.iteration_limit = ITERATION_LIMIT
+        self.game_per_round = GAME_PER_ROUND
 
-    def evaluate_new_score(self):
-        #self.create_test_date()
-        games = self.get_games_played()
+    def calculate_all_games(self, season_name):
+        games = []
+        print ("OK glicko")
```
Comment (blotter) [inline wlrating/glicko2.py]: Ist das eine DEBUG wiedergabe oder wozu sollen die vielen "print" dienen?

File: wlrating/glicko2.py [hunk 13]
```diff
@@ -8,149 +8,233 @@
...
-STANDARD_DEVIATION = Decimal(300).quantize(Decimal('1.00'))
-VOLATILITY = Decimal(0.06).quantize(Decimal('1.00000'))
-VOLATILITY_CHANGE = 0.2
+STARTING_SCORE = 1500
+STANDARD_DEVIATION = 300
+VOLATILITY = 0.06
+TAU = 1
 EPISLON = 0.000001
+ITERATION_LIMIT = 2
+GAME_PER_ROUND = 10
```
Comment (blotter) [inline wlrating/glicko2.py]: In die settings.py?

File: wlrating/admin.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    list_display = ('get_username', 'rating_type', 'decimal1', 'decimal2', 'decimal3', 'season')
+
+    def get_username(self, obj):
+        return obj.user.username
+    get_username.short_description = 'User'
+
+admin.site.register(Player_Rating, PlayerRatingAdmin)
+
+class GameAdmin(admin.ModelAdmin):
+    list_display = ('start_date', 'game_type', 'game_map', 'win_team', 'game_status', 'game_breaks', 'counted_in_score', 'get_participants')
```
[the remark this answers is not in the document]
Comment (Trimardio) [reply wlrating/admin.py]: I'll see if there's an option for that, so it's automatic :P

File: wlrating/view.py (hunk 1 above)
Comment (Trimardio) [reply wlrating/view.py]: Ok, I think I get the idea, but not sure. I should put that in an array, and store it in a file settings.py I create in the module?

File: wlrating/view.py (hunk 2 above)
Comment (Trimardio) [reply wlrating/view.py]: Yeah, sorry about that !

File: wlrating/templates/wlrating/arbiter.html (hunk 4 above)
Comment (Trimardio) [reply wlrating/templates/wlrating/arbiter.html]: done!

File: wlrating/glicko2.py (hunk 12 above)
Comment (Trimardio) [reply wlrating/glicko2.py]: Probably yes!

File: wlrating/glicko2.py (hunk 13 above)
Comment (Trimardio) [reply wlrating/glicko2.py]: See above, where should I put this "settings" file?

**Comment by Trimardio** [comment]
> Maybe create a linked issue for this, and discuss any needed changes in the issue before making a pull request?
> 
> Pull requests on the website don't trigger anything, and we can always get a full diff on GitHub without needing to send of a pull request. One of the advantages over GitHub :)

So I should close this pull request? I'm afraid I'll loose your comments. Or is there an option for that? If I close I don't loose anything right?

File: wlrating/admin.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gunchleoc) [reply wlrating/admin.py]: You can autoformat a file like this:

```
pip install pyformat
pyformat -i <filename>
```

**Comment by gunchleoc** [comment]
You can just keep it open for this one, it's easier then moving the discussion now.

File: wlrating/glicko2.py (hunk 13 above)
Comment (Trimardio) [reply wlrating/glicko2.py]: Ok never mind, found how to, thanks!

File: wlrating/view.py (hunk 1 above)
Comment (blotter) [reply wlrating/view.py]: zum Beispiel so, sieht aber nicht schön aus
```
def type_to_int(word):
    word = word.lower()
    if word == 'autocrat' or word == 'empire':
        return 1
    if word == 'wood gnome' or word == 'barbarian':
        return 2
    if word == 'collectors' or word == 'atlantean':
        return 3
    if word == 'territorial lord' or word == 'frisian':
        return 4
    if word == 'artifacts':
        return 5
```

oder in der settings.py

```
TRIBE = {
    'empire' : 1
    , 'barbarian' : 2
    , 'atlantean' : 3
    , 'frisian' : 4
}
MODI = {
    'autocrat' : 1
    , 'wood gnome' : 2
    , 'collectors' : 3
    , 'territorial lord' : 4
    , 'artifacts' : 5
}
```

gefällt mir irgendwie beides nicht :)

File: wlrating/glicko2.py (hunk 13 above)
Comment (blotter) [reply wlrating/glicko2.py]: in die settings.py unter mainpage/settings.py kommt 

```
GAME_PER_ROUND = 10
```
hier in der datei dann

```
from django.conf import settings

[...]
self.game_per_round = settings.GAME_PER_ROUND
[...]
```

oder

```
from django.conf import settings

[...]
def get(key, default):
    return getattr(settings, key, default)
[...]
self.game_per_round = get('GAME_PER_ROUND', 10)
[...]
```

**Comment by Trimardio** [comment]
@frankystone I know you have less time but just about the tribe model I have a few questions:

* There are currently no admin page, I could modify directly the db, but I guess for the management of the rating system it's better if it can be modified easily with the rest. Notably for tests. So I should I create one? Or should I put the admin page for the tribe in the wlrating module?

* I also need a table for type of games and maps, should I put those in the wlhelp page because they might be needed in the wiki too?

Other refacto questions for everyone:

* I have a "test" function that automatically create games to test if the glicko system works fine. Should I definitly delete those or is there some place for test code?

* The view page is getting pretty big, is there a moment where it makes sense in separating the views in many pages? 

* I have currently a class called "Temporary_user". I need it to know which user is participating in the rating. Should I keep it? Or should I rather add a property to the user model (participate_in_rating boolean)?

**Comment by frankystone** [comment]
> @frankystone I know you have less time but just about the tribe model I have a few questions:
> 
> * There are currently no admin page, I could modify directly the db, but I guess for the management of the rating system it's better if it can be modified easily with the rest. Notably for tests. So I should I create one? Or should I put the admin page for the tribe in the wlrating module?

I guessed you need it just to get a valid Tribename, or get a dropdown field to choose a tribe. If you need more than this you have to add an additional table, because the tribes table in the encyclopedia is sticked with the help.

> * I also need a table for type of games and maps, should I put those in the wlhelp page because they might be needed in the wiki too?

In general i think the rating module should be a standalone module, like it is with all other modules. Ideally it should be detached from widelands so it can be used by other projects too, although it may be too widelands specific.

So create the needed tables only in your module.

> 
> Other refacto questions for everyone:
> 
> * I have a "test" function that automatically create games to test if the glicko system works fine. Should I definitly delete those or is there some place for test code?

Normally each module 'should' have some tests. I missed such tests when switching to python3, but i found no time to write them. So it's a very good idea to write tests :-)

See https://docs.djangoproject.com/en/1.11/intro/tutorial05/

> * The view page is getting pretty big, is there a moment where it makes sense in separating the views in many pages?

No, i think this depends on personal preferences and also to the size of the used screen.

> * I have currently a class called "Temporary_user". I need it to know which user is participating in the rating. Should I keep it? Or should I rather add a property to the user model (participate_in_rating boolean)?

The user model is part of django, so you can't edit it ;) We have extended this model in wlprofile, but as said above: Your module should keep all it needs in itself.

File: wlrating/view.py (hunk 1 above)
Comment (Trimardio) [reply wlrating/view.py]: I just created a whole model for that use instead :)

**Comment by Trimardio** [comment]
Mhhh, I'm tryoing to update to last version of the website, but now I get this error when trying to update. Is there some new things that have to be done for models to work fine?

"RuntimeError: Model class wlrating.models.Season doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS."

**Comment by Trimardio** [comment]
Huh, I tried to update but now I get this new error:

"AttributeError: 'Settings' object has no attribute 'NORECAPTCHA_SITE_KEY'" 

When trying to migrate.

Ideas?

**Comment by frankystone** [comment]
Probably you didn't have local_settings.py copied into the directory mainpage?

**Comment by Trimardio** [comment]
> Probably you didn't have local_settings.py copied into the directory mainpage?

Nope, I did already :(

**Comment by stonerl** [comment]
@Trimardio 

https://developers.google.com/recaptcha/docs/faq#id-like-to-run-automated-tests-with-recaptcha-v2-what-should-i-do

**Comment by Trimardio** [comment]
@stonerl I did at it says there:

`NORECAPTCHA_SITE_KEY = '6LeIxAcTAAAAAJcZVRqyHh71UMIEGNQ_MXjiZKhI'
NORECAPTCHA_SECRET_KEY = '6LeIxAcTAAAAAGG-vFI1TnRWxMZNFuojJ4WifJWe'`

Still no luck :( And I read elswhere the 'dummy' string should work anyway. Maybe I made something weird elsewhere? I have to say, I have the same problem with the official repo untouched.

**Comment by stonerl** [comment]
Maybe commenting this line out could help.
https://github.com/widelands/widelands-website/blob/51754f1b6ec812b880e375a56f6becbd13f7c867/mainpage/forms.py#L13

**Comment by Trimardio** [comment]
> Maybe commenting this line out could help.
> 
> https://github.com/widelands/widelands-website/blob/51754f1b6ec812b880e375a56f6becbd13f7c867/mainpage/forms.py#L13

Ok! So that DID work (after commenting, and adding a "pass", if anyone has the same problem)! Thank a lot! :D 

But, doesn't that mean something should be change in the code then? Or at least in the readme? I certainly won't be the only one to have this problem. Unless it's specific to my environment, but I wouldn't see why. I use a wrapper.

**Comment by frankystone** [comment]
> Huh, I tried to update but now I get this new error:
> 
> "AttributeError: 'Settings' object has no attribute 'NORECAPTCHA_SITE_KEY'"

This error says Django can't find this setting. IMHO this has nothing to do with the captcha form.

I have setup your branch by downloading a Zipped version, setup the virtualenvironment from scratch and could not reproduce the error you have.

Please make sure the file 'mainpage/local_settings.py' exists, and contains the lines:
`NORECAPTCHA_SITE_KEY = 'dummy'`
`NORECAPTCHA_SECRET_KEY = 'dummy'`
Or you can do this in a django management console:
start the console by executing `$> ./manage.py shell`
In there type:

`from django.conf import settings`
`settings.NO`   # Press the TAB key twice
This will show every setting starting with NO...:
`settings.NORECAPTCHA_SECRET_KEY  settings.NORECAPTCHA_SITE_KEY   ...`

**Comment by Trimardio** [comment]
Nice command.

Sorryyyyyy, I was so sure it was stupid. I read you without reading, just like I read the readme without reading it. My god. The mainpage directory. Just didn't get that. Was so sure of myself.

Ok, for not confusing anyone else after me, I think local_settings.sample should be in mainpage too. I'll write a bug report for other confusing thing in the readme in the same occasion.

**Comment by Trimardio** [comment]
As said in the forum, I think this version is ready for the alpha. It would help for me to get more input, and other people to actually see how that works. I've been so long on that, I think there are obvious problem I'm not seeing.

If alpha is too complicated to implement I would understand, but from what I understood, it didn't seem to take too much time? No need to merge for the test,  I think you guys can directly pull from my repo if I understand correctly.

**Comment by stonerl** [comment]
@Trimardio would you be so kind and merge the latest master into your branch?

**Comment by stonerl** [comment]
@Trimardio would you be so kind and resolve the conflicts?

**Comment by Trimardio** [comment]
> 
> 
> @Trimardio would you be so kind and resolve the conflicts?

So sorry, this will take time, I'm getting a lot of work IRL for studies. But after that I'll gladly do it, I miss so much working on it!
I guess this pull will be on pause for some time.

**Comment by SirVer** [comment]
I am going to close this one. It would have been such a great feature, but in the last 2 years this has bit rotted, so merging this will be substantial work. @Trimardio unless you are up to, if you are let me know and we can reopen.

**Pull request closed without merging**